netx/dialer/quicsaver.go

```diff
@@ -0,0 +1,126 @@
+package dialer
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"time"
+
+	"github.com/lucas-clemente/quic-go"
+	"github.com/ooni/probe-engine/internal/tlsx"
+	"github.com/ooni/probe-engine/netx/errorx"
+	"github.com/ooni/probe-engine/netx/trace"
+)
+
+// QUICSaverDialer saves events occurring during the dial
+type QUICSaverDialer struct {
+	QUICContextDialer
+	Saver *trace.Saver
+}
+
+// DialContext implements Dialer.DialContext
+func (d QUICSaverDialer) DialContext(ctx context.Context, network, addr string, host string, tlsCfg *tls.Config, cfg *quic.Config) (quic.EarlySession, error) {
+	start := time.Now()
+	sess, err := d.QUICContextDialer.DialContext(ctx, network, addr, host, tlsCfg, cfg)
+	stop := time.Now()
+	d.Saver.Write(trace.Event{
+		Address:  host,
+		Duration: stop.Sub(start),
+		Err:      err,
+		Name:     errorx.QUICHandshakeOperation,
+		Proto:    network,
+		Time:     stop,
+	})
+	return sess, err
+}
+
+// QUICHandshakeSaver saves events occurring during the handshake
+type QUICHandshakeSaver struct {
+	Saver  *trace.Saver
+	Dialer QUICContextDialer
+}
+
+// DialContext implements QUICContextDialer.DialContext
+func (h QUICHandshakeSaver) DialContext(ctx context.Context, network string, addr string, host string, tlsCfg *tls.Config, cfg *quic.Config) (quic.EarlySession, error) {
+	start := time.Now()
+	h.Saver.Write(trace.Event{
+		Name:          "quic_handshake_start",
+		NoTLSVerify:   tlsCfg.InsecureSkipVerify,
+		TLSNextProtos: tlsCfg.NextProtos,
+		TLSServerName: tlsCfg.ServerName,
+		Time:          start,
+	})
+	sess, err := h.Dialer.DialContext(ctx, network, addr, host, tlsCfg, cfg)
+	stop := time.Now()
+
+	if sess == nil {
```

Comment:
I feel like this is better written as

```suggestion
	if err != nil {
```